src/Network/Transport/TCP.hs

```diff
@@ -828,17 +834,28 @@ apiCloseEndPoint transport evs ourEndPoint =
             putMVar resolved ()
             return closed
           RemoteEndPointValid vst -> do
+            -- Schedule an action to send a CloseEndPoint message and then
+            -- wait for the socket to actually close (meaning that this
+            -- end point is no longer receiving from it).
+            -- Since we replace the state in this MVar with 'closed', it's
+            -- guaranteed that no other actions will be scheduled after this
+            -- one.
             sched theirEndPoint $ do
               void $ tryIO $ sendOn vst [ encodeInt32 CloseEndPoint ]
               -- Release probing resources if probing.
               forM_ (remoteProbing vst) id
-              tryCloseSocket (remoteSocket vst)
```

Comment:
Good catch. This relies on the peer behaving well and closing their socket when `CloseEndPoint` is received. I'll patch this so that it will shutdown the socket and I'll make a test case to verify that it works.